### ISSUE TITLE
Unify the convention of functions that operate on multiple values

### ISSUE THREAD

**Issue by xiaq** [elves/elvish#991]
Elvish has two different conventions for functions that operate on arbitrarily many values:

* **Variadic** functions take an arbitrary amount of arguments (possibly after several fixed arguments). Examples include `echo`, `pprint`, `+`.

* **argument-or-input** functions take an optional list argument. When the list argument is present, it operates on the values in the list. When the list argument is missing, it operates on the values from the input. Examples include `each`, `to-json`, `str:join`.

It is not always clear which category a function falls into or should fall into, and it can get confusing when people make the wrong assumption. Some examples:

* It is confusing that `to-json []` outputs nothing instead of `'[]'`. The correct way to serialize `[]` into JSON is `put [] | to-json`, or `to-json [[]]`, since `to-json` is an argument-or-input function.

* While `pprint` is a variadic argument, there is a request to make it take input values (#736). The current solution is `... | pprint (all)`. That issue also contains some earlier discussions about this topic.

* It is not clear which category the to-be-introduced `math:min` and `math:max` functions should be in (#967).

The obvious way to unify these two conventions is to let commands (1) be variadic, and (2) read from inputs when there are no arguments. Indeed, this is the convention used by almost all the classical UNIX text-processing commands. The problem with that is with code like this:

```sh
cat $@files
```

And it can be surprising if `$files` is an empty list. The programmer might have intended this code to read from stdin when `$files` is empty, but it's more likely that they intended this code to not do anything while `$files` is empty. Since there are many constructs in Elvish that output multiple values (output capture for example), this problem can be very pervasive.

We cannot fix problems in the classical tools, but at least we can avoid the same problem in Elvish's builtin functions. In fact, the motivation of the **argument-or-input** convention was to avoid this mistake. Consider the following examples with `each`:

```
each [x]{ cat $x } $files
```

When `$files` is empty, `each` gets an empty list, so it's clear that it should do nothing, instead of reading from input.

**Comment by xiaq** [comment]
This is not a new problem: any language that support both variadic arguments has to choose whether each function should accept variadic arguments or a list, and it's sometimes arbitrary. In Elvish the problem gets more complex since the problem of inputs gets involved.

The `max` function in Python accepts both variadic arguments, or a list:

```python
max(1, 2, 3) # this works
max([1, 2, 3]) # this also works
```

(I do not consider this an example to be emulated.)

**Comment by krader1961** [comment]
> argument-or-input functions take an optional list argument....

That is an incomplete description of the current behavior since "argument-or-input" style commands also take zero or more fixed arguments just like variadic functions. Which you included the in characterization of "variadic" style commands. Consider the builtin `joins` command which has one fixed argument, the separator string, that must be present.

**Comment by hanche** [comment]
One tiny bit of anecdata: I got bitten by the `to-json` behaviour just the other day.

**Comment by krader1961** [comment]
With regard to this example:

```
cat $@files
```

@xiaq wrote:

> And it can be surprising if $files is an empty list. The programmer might have intended this code to read from stdin when $files is empty, but it's more likely that they intended this code to not do anything while $files is empty.

I agree it can be surprising if `$files` is an empty list that expands to zero arguments. If there is no elvish function named `cat` I don't see how to avoid the problem. If `cat` is an external command it is run with zero arguments and reads from stdin. If `cat` is an elvish function that is a different situation.

Wouldn't it be reasonable to communicate to the function whether or not there was an argument not tied to a fixed parameter even if that argument expanded to zero values? That would allow the function, and the generic "argument-or-input" processing code, to fallback on reading from their stdin stream only if no arguments were supplied; regardless of the fact they expanded to zero values. Or do something else that was appropriate for the situation.

In other words, if I `cat $@files` and `$files` is an empty list I don't want an elvish `cat` to read from stdin. The question is how to square that behavior with that of external commands that are run when a var expansion is empty.

**Comment by xiaq** [comment]
@krader1961 re the description of argument-of-input functions, I corrected it :)

@krader1961 re "sending information on whether there was an argument": that works in the limited example you gave, but consider this immediate difficulty: say `str:join` is converted to such a function; `str:join ',' $@a` works, but what about `str:join $@a`? Was there an "intended final argument" or not - does it depend on the length of `$a`?

But that is just an illustrative example, the more nasty issue is that it breaks a basic, almost tautological, principle in programming languages: if expression `X` and expression `Y` evaluates to the same, they should behave the same in all contexts. (I am not aware of a name for this principle, but let's call this the *equivalency principle*.) In Elvish's case, `$@a` evaluates to *no values* if `$a` is empty, so it must behave the same with no values; breaking this will only lead to more difficult corner cases (I'm quite sure one can find such more over time).

One may argue that use of `str:join $@a` must be really rare because the first argument (the separator) and the rest (the strings to join) are quite different and shouldn't be mixed in one `$a` list. While this still does not justify breaking the equivalency principle, this does correctly observe that "variadic arguments" (a term I now use liberally for any conceptually arbitrary many arguments) are usually quite different from "fixed arguments" and they usually do not mix. In fact, the convention of argument-or-input function is a manifest of that: the "variadic arguments" in this case are either supplied as inputs (completely separate from fixed arguments) or a list (visually separate from fixed arguments). Even in the case of true variadic functions, almost all of them do not take any fixed arguments at all, so there are only variadic arguments.

---

**Possible solution 1**.

The observation leads to one hypothetical design: make a syntactical distinction between "variadic arguments" from normal arguments. This is comparable to how Elvish makes a syntactical distinction between options and arguments. Let's say we make `%` a metacharacter for separating fixed arguments and variadic arguments (this is clearly just for illustration's purpose), the unified convention works like this:

* Variadic arguments can only appear after the separator. There could be 0 variadic argument.

* If the seprator appears syntactically in the command, the command operatees on those arguments.

* If the seprator does not appear syntactically in the command, the command operates on values from the input.

Examples:

```sh
str:join ',' % a b c # writes "a,b,c"
str:join ',' % # writes ""
str:join ',' # reads from input
put a b c | str:join ',' # writes "a,b,c"
str:join ',' % $@a # joins elements in $a
```

This syntax is quite acceptable for current argument-or-input functions, but it makes the current variadic functions much harder to use. Having to write `+ % 1 2 3` is exceedingly annoying.

---

**Possible solution 2**.

Another idea is to make the "read from input" behavior explicit rather than implicit. So the rules are:

* Functions take variadic arguments.
* If there is a placeholder "read from input" argument, say $input, the function reads from the input.

This does not change anything for the current variadic arguments. For current argument-or-input functions, it works like this (still using `str:join` as an example):

```sh
str:join ',' a b c # writes "a,b,c"
str:join ',' # writes ""
str:join ',' $input # reads from input
put a b c | str:join ',' $input # reads from input
```

This is actually kind of nice, but my problem with this is that it makes the style of passing values via the pipeline more verbose, and discourages it. Users who do not use the pipeline style a lot may find this fine; but I have a strong preference to encourage *more* pipeline uses, not fewer.

---

**Non-solution; an idea**.

In #736 I mentioned an idea of making argument-or-input functions only take values from the input, never arguments. Using `str:join` as an example again, disallow this:

```sh
str:join ',' [a b c]
```

And only allow this:

```sh
put a b c | str:join ','
```

And to make this *slightly* more ergonomic, make lists write out their values when called, so that the following also does the same thing:

```sh
[a b c] | str:join ','
```

This is an idea motivated by "pipeline fundamentalism" and does not solve the problem in this issue, but perhaps some variant of this can be part of the overall solution.

**Comment by xiaq** [comment]
To clarify, I like neither of two "possible solutions" I put forward above, they are intended as straw man proposals.

**Comment by krader1961** [comment]
I need to think more about this problem but I'll point out that the fact the legacy `joins`, and the recently added `str:join`, require a single, fixed, separator argument is arguably a mistake. It looks like this is an artifact, presumably, of `builtin:joins` having been written before function options were supported. Today it would probably be better for both commands to take a `&sep` option that defaults to a single space or an empty string. That would address a couple of issues raised in this issue. Such as what `str:join $@list` means. My opinion is it means the first element of the list is the separator char. But if the separator has to be specified as an option then that is no longer a question.

**Comment by zzamboni** [comment]
@krader1961 my guess is that the mandatory separator argument is because historically a lot of "join" functions require one. In a cursory search, the "join" functions in Perl, Python and Go all require a separator, whereas Clojure's and PHP's default to an empty string.

@xiaq as for the main topic, I think both the variadic and argument-or-input make sense, and which one makes most sense depends on the function. At the moment only variadic is a language feature, since argument-or-input parameters need to be implemented by each function (I even wrote such a function [`util:optional-input`](https://github.com/zzamboni/elvish-modules/blob/master/util.org#pipeline-or-argument-input)).

How about leaving still the choice of the model to each function, but make argument-or-input also a language feature? For example, just like variadic arguments are declared with `@args`, argument-or-input could be declared with `%args`.

**Comment by rdw20170120** [comment]
Solving this problem is one of the primary goals of my own open-source project.  I suggest that this is NOT a _**language problem**_, but rather a _**usage problem**_.  I think that it is critically important to distinguish between the two.  I think that it is also critically important to correctly identify which category fits any particular use case.  Allow me to make this a bit more concrete.

If I say `cat $File`, and `File` is undefined, I love that Elvish treats this as a _**language problem**_ and rightly throws an exception.  I hate that other scripting languages do not handle this, or make it extremely difficult to do so consistently.

If I am referencing the variable `File` in my script, but defining it is not within the scope of my own code (for whatever reason), then I might have a problem.  How do I avoid throwing the exception, if my code got called but the caller did not define `File`?  That is a _**usage problem**_, and is NOT a _**language problem**_.  I can simply write a test for the existence of the variable `File` before I invoke `cat $File`.  That rightly shifts to problem to the script writer instead of putting it upon Elvish.

Assuming that I have written such a conditional to test for the existence of somebody else's variable `File`, what else can go wrong?  As discussed, what if `$File` is empty (in its various forms)?  Again, I classify this as a _**usage problem**_, NOT a _**language problem**_.  It is NOT Elvish's problem to protect the user from every possible way that they might misuse the various syntax and commands based on trivial things like whether they actually provided a non-empty argument.

I think that it is essential for us to notice that _**no programming language can EVER have enough information to properly guess**_ how to handle such a situation.  We are not mind readers, and neither is the machine running our programming language.  Therefore, we should NOT try.  Instead, we should clearly categorize this as a _**usage problem**_, then ensure that we have provided the user with some recourse for handling such issues in the various ways that MIGHT apply in any particular situation.

I perceive the discussion above as perhaps being stuck on such an attempt to _**guess**_ how to properly handle an unknown _**usage problem**_; I see it as "borrowing trouble".  I think that much of what we end up hating about various programming languages can be attributed to getting this wrong:  attempting to treat a _**usage problem**_ as a _**language problem**_, thereby making the language arbitrary and ugly in some particular aspect (such as the comment above about Python's `max`).

I have already implemented part of my project's solution to this _**usage problem**_ in Fish.  I intend to do the same in Elvish as I make the transition.

I wrote a series of Fish functions that I (and my users) can invoke to routinely solve these kinds of _**usage problems**_, which are rampant in SRE/DevOps use cases.  I NEVER write `cat $File`, then let it go bad in all the various ways when a user executes the script.  Instead, (in Fish) I write `cat (file (var File))`.  In this case, `cat` is still the external command (or perhaps a wrapper) that will still fail if used improperly.  `File` is still a reference to a variable that might not be defined yet, which is NOT protected in Fish as it is in Elvish.  Both `file` and `var` are Fish functions that I wrote to handle this _**usage problem**_.

My `file` and `var` functions, and the rest of their siblings, allow me to apply a very simple pattern to protect my script.  These functions are called using output-capture `()` syntax, which is the same in Elvish and Fish for this use case.  Each function takes a single input, and returns a single output.  Along the way, each function asserts various conditions (side effects) that must be true at that point in the calling script regarding the provided input.  If any condition fails, an exception terminates the script (fail-fast) with a usable error message pointing directly at the location in the script.

In a few cases such as `var`, the output is a simple and obvious transformation of the input.  My `var` function expects the name of a variable and it outputs its value (if any).  If the variable is undefined, then `var` throws an exception.

In most other cases such as `file`, the output is (nearly) identical to the input.  This function validates that the input is a single (not empty, not variadic) value that refers to an existing regular file that is readable by the user.  Along the way, the value is transformed into the canonical absolute pathname of the file.

Going back to `cat (file (var File))`, we can now see that this is a very concise but readable invocation to output the contents of `File`, BUT _**it handles the unhappy paths in a fail-fast manner with clear messages**_.  This simple invocation protects the script against ALL these unhappy paths, while giving a stack trace with these (almost literal) error messages:
- `variable "File" is UNDEFINED`
- `Value is EMPTY`
- `Value is MULTIPLE instead of singular`
- `Path "PATH" does NOT EXIST`
- `Path "PATH" is NOT A REGULAR FILE`
- `File "FILE" is NOT READABLE`

As you can see, it is then easy to compose these little validation functions to cover a wide variety of _**usage problems**_ that arise in scripts.  These functions properly handle these problems instead of burdening the relevant language to do so.  I believe that this is a crucial concept that needs to be communicated and practiced across our industry, especially when it comes to infrastructure automation.  Solving this ONE problem will have a massive positive impact.  I hope to help solve many others along the way with appropriate understanding, patterns, and simple implementation.

I think that matches and complements what is being attempted with Elvish as well.  Thank you!

Back to the original problem at hand...  I suggest that Elvish should not try to guess, but instead should require the user to be explicit about their intent.  I suggest that Elvish should provide a simple way in the syntax for the user to tell any command that it should expect its input on STDIN rather than as arguments.  I wonder whether the common convention of a `-` argument would serve that purpose.  If so, could Elvish then apply that consistently, and also make it easy for users to apply the same convention to their own functions?

What do you think?

**Comment by rdw20170120** [comment]
By the way, in Elvish I am anticipating a syntax like `cat (f (v File))` for the same purpose as what I wrote above for Fish.  I feel that these little validation functions will be so omnipresent that they deserve single-letter abbreviations.  Elvish already handles undefined variables (yeah!), nevertheless, I will still write something like my `var` function to handle cases where the variable is defined outside the script author's scope.
